pkg/core/dao/dao.go

```diff
@@ -300,30 +300,34 @@ func (dao *Simple) GetStorageItemsWithPrefix(id int32, prefix []byte) ([]state.S
 			Item: state.StorageItem(v),
 		})
 	}
-	dao.Seek(id, prefix, saveToArr)
+	dao.Seek(id, storage.SeekRange{Prefix: prefix}, saveToArr)
 	return siArr, nil
 }
 
-// Seek executes f for all items with a given prefix.
-// If key is to be used outside of f, they may not be copied.
-func (dao *Simple) Seek(id int32, prefix []byte, f func(k, v []byte)) {
-	lookupKey := makeStorageItemKey(dao.Version.StoragePrefix, id, nil)
-	if prefix != nil {
-		lookupKey = append(lookupKey, prefix...)
+// Seek executes f for all storage items matching a given `rng` (matching given prefix and
+// starting from the point specified). If key or value is to be used outside of f, they
+// may not be copied.
+func (dao *Simple) Seek(id int32, rng storage.SeekRange, f func(k, v []byte)) {
+	prefixLen := len(rng.Prefix)
+	rng.Prefix = makeStorageItemKey(dao.Version.StoragePrefix, id, rng.Prefix)
+	if len(rng.Start) != 0 {
```

Comment:
> will overwrite the old prefix value

Nice catch, it can be fixed either by an extra check inside dao.Seek or by allocating a separate slice for `rng.Start`, currently, the second option is implemented. Although it's a program error for `rng.Start` not to be prefixed by `rng.Prefix`.

>  if Start was defined as a value appended to prefix

I also thought about it, see https://github.com/nspcc-dev/neo-go/pull/2316#discussion_r782794072. I think it's a nice solution to avoid extra checks. Implement?

Reply:
Doesn't look bad to me, avoids some checks and potential errors.

Reply:
Implemented, and it actually simpler than I expected. And the code looks cleaner.